### PR TITLE
[QAOA (Secondary)] Adds Cost Hamiltonian Components

### PR DESCRIPTION
**Pull request by Lucaman99**
**Context:**

Part of the ongoing effort to add QAOA functionality to PennyLane

**Description of the Change:**

- Adds "cost Hamiltonian components", which are smaller "sub-Hamiltonians" that can be used to construct specific cost Hamiltonians, corresponding to different optimization problems.
- Updates the built-in `qaoa.maxcut` method to take advantage of these new "cost components".

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qaoa/cost.py
Comment (trbromley) [inline pennylane/qaoa/cost.py:211]: Minor thing: wouldn't it be like:
```pycon
>>> print
result
>>> print
result
```

File: pennylane/qaoa/cost.py
Comment (trbromley) [inline pennylane/qaoa/cost.py:222]: Was the approach before incorrect? It looks like more than just changing the method, but the actual result is different?

Comment (trbromley) [reply pennylane/qaoa/cost.py:222]: oh maybe just the identities were different :thinking:

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @Lucaman99, great job with the code as always, just a couple suggestions on the docstrings.

File: pennylane/qaoa/cost.py
Comment (Lucaman99) [reply pennylane/qaoa/cost.py:222]: @trbromley Yes, the only difference is the identities, which get added together (as the new approach to constructing this Hamiltonian involves addition of terms, which automatically simplifies the newly-created Hamiltonian)

**Comment by Lucaman99** [comment]
@trbromley @ixfoduap Hopefully the most recent changes make the docstring a bit clearer!

**Review by trbromley — APPROVED** [review]
Thanks @Lucaman99!

**Comment by josh146** [comment]
Likewise, I presume we're just waiting on @ixfoduap's review for this PR before it can be merged in?

**Comment by Lucaman99** [comment]
@josh146 Yes!

File: pennylane/qaoa/cost.py
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:95]: I'm still confused with the definition of this Hamiltonian, even after looking at the example.

1. What does it mean for an edge to be assigned an energy? _States_ (actually eigenstates of the Hamiltonian) can be associated with an energy, but the edges just denote a pair of qubits.

2. I don't think that the connection between the Hamiltonian and the reward is too clear. Maybe good to add in text below the code snippet what the expectation values are for some states? Something like
"For example, a quick calculation gives the following expectation values:
<000|H|000> = 1 (I think)
<110|H|110> = -0.25 (I think)
meaning, a colouring that assigns "00" to the first two qubits has higher energy than a colouring that assigns them with "11" "

File: pennylane/qaoa/cost.py
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:112]:
```suggestion
        For example, given the reward set: :math:`\{|00\rangle, \ |01\rangle, \ |10\rangle\}` and the graph :math:`G`,
```

File: pennylane/qaoa/cost.py
Comment (ixfoduap) [reply pennylane/qaoa/cost.py:95]: The usage details are helpful, but I think the explanation beforehand can be made clearer

File: pennylane/qaoa/cost.py
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:125]:
```suggestion
            Within an undirected graph, there is no notion of "order"
```

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Looks solid! Left a few comments regarding clarity of the docstrings

File: pennylane/qaoa/cost.py
Comment (Lucaman99) [reply pennylane/qaoa/cost.py]: Hopefully it's more clear now!

**Review by ixfoduap — APPROVED** [review]

**Pull request merged**
